docs/en/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,869 @@
+
+The following is an updated `MyEncapsulatedStateWidget` example that will correctly update its output when its state is changed.
+
+**Function-based variant:**
+
+```tsx
+import { create, tsx, invalidator } from '@dojo/framework/core/vdom';
+
+const factory = create({ invalidator });
+
+let myState: string = 'Hello from a stateful widget!';
+let counter: number = 0;
+
+export default factory(function MyEncapsulatedStateWidget({ middleware: { invalidator } }) {
+	return (
+		<div>
+			Current widget state: {myState}
+			<br />
+			<button
+				onclick={() => {
+					myState = 'State change iteration #' + ++counter;
+					invalidator();
+				}}
+			>
+				Change State
+			</button>
+		</div>
+	);
+});
+```
+
+**Class-based variant:**
+
+Here, both `myState` and `counter` are updated as part of the same application logic operation, so `@watch()` could be added to either or both of the fields, with the same net effect and performance profile in all cases:
+
+> src/widgets/MyEncapsulatedStateWidget.tsx
+
+```tsx
+import WidgetBase from '@dojo/framework/core/WidgetBase';
+import watch from '@dojo/framework/core/decorators/watch';
+import { tsx } from '@dojo/framework/core/vdom';
+
+export default class MyEncapsulatedStateWidget extends WidgetBase {
+	private myState: string = 'Hello from a stateful widget!';
+
+	@watch() private counter: number = 0;
+
+	protected render() {
+		return (
+			<div>
+				Current widget state: {this.myState}
+				<br />
+				<button
+					onclick={() => {
+						this.myState = 'State change iteration #' + ++this.counter;
+					}}
+				>
+					Change State
+				</button>
+			</div>
+		);
+	}
+}
+```
+
+## Intermediate: passing widget properties
+
+Passing state into a widget via virtual node [`properties`](#node-properties) is the most effective way of wiring up reactive data flows within a Dojo application.
+
+Widgets specify their own properties [interface](https://www.typescriptlang.org/docs/handbook/interfaces.html) which can include any fields the widget wants to publicly advertise to consumers, including configuration options, fields representing injectable state, as well as any event handler functions.
+
+Function-based widgets pass their properties interface as a generic type argument to the `create().properties()` call. The factory returned from this call chain then makes property values available via a `properties` argument in the render function definition.
+
+Class-based widgets can define their properties interface as a generic type argument to `WidgetBase` in their class definition.
+
+For example, a widget supporting state and event handler properties:
+
+> src/widgets/MyWidget.tsx
+
+**Function-based variant:**
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+
+const factory = create().properties<{
+	name: string;
+	onNameChange?(newName: string): void;
+}>();
+
+export default factory(function MyWidget({ properties: { name, onNameChange } }) {
```

Comment:
Similar comment to below, the `newName` value needs to persist across renders and be passed into the `input` here, something like this:

```ts
const factory = create({ icache }).properties<{
	name: string;
	onNameChange?(newName: string): void;
}>();

export default factory(function MyWidget({ middleware: { icache }, properties: { name, onNameChange } }) {
	let newName = cache.get<string>('new-name') || '';
	return (
		<div>
			<span>Hello, {name}! Not you? Set your name:</span>
			<input
				type="text"
				value={newName}
				oninput={(e: Event) => {
					icache.set('new-name', (e.target as HTMLInputElement).value);
				}}
			/>
			<button
				onclick={() => {
					icache.set('new-name', undefined);
					onNameChange && onNameChange(newName);
				}}
			>
				Set new name
			</button>
		</div>
	);
});
```